sdk/python/lib/pulumi/runtime/resource.py

```diff
@@ -97,6 +96,104 @@ class ResourceResolverOperations(NamedTuple):
     """
 
 
+def create_urn(
+    name: "Input[str]",
+    type_: "Input[str]",
+    parent: Optional["Input[str]"] = None,
+    project: Optional[str] = None,
+    stack: Optional[str] = None,
+) -> "Output[str]":
+    """
+    create_urn computes a URN from the combination of a resource name, resource type, optional
+    parent, optional project and optional stack.
+    """
+    parent_prefix: Optional[Output[str]] = None
+    if parent is not None:
+        parent_urn = Output.from_input(parent)
+        parent_prefix = parent_urn.apply(lambda u: u[0 : u.rfind("::")] + "$")
+    else:
+        if stack is None:
+            stack = settings.get_stack()
+
+        if project is None:
+            project = settings.get_project()
+
+        parent_prefix = Output.from_input("urn:pulumi:" + stack + "::" + project + "::")
+
+    all_args = [parent_prefix, type_, name]
+    # invariant http://mypy.readthedocs.io/en/latest/common_issues.html#variance
+    return Output.all(*all_args).apply(lambda arr: arr[0] + arr[1] + "::" + arr[2])  # type: ignore
+
+
+def collapse_alias_to_urn(
+    alias: "Input[Union[Alias, str]]",
+    defaultName: str,
+    defaultType: str,
+    defaultParent: Optional["Resource"],
+) -> "Output[str]":
+    """
+    collapse_alias_to_urn turns an Alias into a URN given a set of default data
+    """
+
+    # pylint: disable-next=used-before-assignment
+    def collapse_alias_to_urn_worker(inner: Union[Alias, str]) -> Output[str]:
+        if isinstance(inner, str):
+            return Output.from_input(inner)
+
+        name = inner.name if inner.name is not ... else defaultName  # type: ignore
+        type_ = inner.type_ if inner.type_ is not ... else defaultType  # type: ignore
+        parent = inner.parent if inner.parent is not ... else defaultParent  # type: ignore
+        project: str = inner.project if inner.project is not ... else settings.get_project()  # type: ignore
+        stack: str = inner.stack if inner.stack is not ... else settings.get_stack()  # type: ignore
+
+        if name is None:
+            raise Exception("No valid 'name' passed in for alias.")
+
+        if type_ is None:
+            raise Exception("No valid 'type_' passed in for alias.")
+
+        parent_urn: Optional["Input[str]"] = None
+        if parent is not None:
+            # pylint: disable-next=used-before-assignment
+            if isinstance(parent, Resource):
+                parent_urn = parent.urn
+            else:
+                parent_urn = parent
+        return create_urn(name, type_, parent_urn, project, stack)
+
+    inputAlias: Output[Union[Alias, str]] = Output.from_input(alias)
+    return inputAlias.apply(collapse_alias_to_urn_worker)
+
+
+async def create_alias_spec(resolved_alias: Alias) -> alias_pb2.Alias.Spec:
+    alias_spec = alias_pb2.Alias.Spec()
+    if resolved_alias.name is not None:
+        alias_spec.name = resolved_alias.name
+    if resolved_alias.type_ is not None:
```

Comment:
Why is ours `type_` but protobufs just `type`?

Reply:
was exactly my thought 😅